usr.sbin/asf/asf.8
.\" Copyright (c) 2003 Greg Lehey.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" This software is provided by Greg Lehey ``as is'' and
.\" any express or implied warranties, including, but not limited to, the
.\" implied warranties of merchantability and fitness for a particular purpose
.\" are disclaimed.  in no event shall Greg Lehey be liable
.\" for any direct, indirect, incidental, special, exemplary, or consequential
.\" damages (including, but not limited to, procurement of substitute goods
.\" or services; loss of use, data, or profits; or business interruption)
.\" however caused and on any theory of liability, whether in contract, strict
.\" liability, or tort (including negligence or otherwise) arising in any way
.\" out of the use of this software, even if advised of the possibility of
.\" such damage.
.\"
.\" $FreeBSD: stable/11/usr.sbin/asf/asf.8 335518 2018-06-22 00:27:31Z emaste $
.\"
.Dd December 20, 2006
.Dt ASF 8
.Os
.Sh NAME
.Nm asf
.Nd add symbol files
.Sh SYNOPSIS
.Nm
.Op Fl afKksVx
.Op Fl M Ar core
.Op Fl N Ar system
.Op Fl o Ar outfile
.Op Fl X Ar suffix
.Op Ar modules-path Op Ar outfile
.Sh DEPRECATION NOTICE
The
.Nm
utility will be removed in a future release.
The
.Xr kgdb 1
debugger handles kernel module state internally.
.Sh DESCRIPTION
By default,
.Nm
reads
.Xr kldstat 8
output from standard input and writes to the
.Pa .asf
file a list of
.Xr gdb 1
commands to add symbol files from KLDs in subdirectories of the subdirectory
.Pa modules
of the current directory, which is intended to be a kernel build directory.
This allows
.Xr gdb 1
to load the symbols into the debugging environment.
.Pp
An optional
.Ar modules-path
argument can specify a semicolon-separated list of directory pathnames
similar to the
.Va kern.module_path
sysctl.
Each directory in the list will be searched in turn for modules.
The default list consists of just one element,
.Pa modules ,
which is suitable if the current directory is a kernel build directory.
.Pp
If
.Ar outfile
is specified,
.Nm
writes to it instead of
.Pa .asf .
If
.Ar outfile
is a single dash
.Pq Sq Fl ,
standard output is used.
.Sh OPTIONS
The following options modify the function of
.Nm :
.Bl -tag -width indent
.It Fl a
When writing to an explicit
.Ar outfile ,
append to the file rather than overwriting it.
.It Fl f
Instead of trying to simplistically guess the path for each module, perform
a traversal in the same way that
.Xr find 1
does to locate an exact path for each module, no matter where in
.Ar modules-path
it is located.
.It Fl K
Instead of reading from standard input, use the conventional
system interface to get the list of modules currently loaded.
.It Fl k
Instead of reading from standard input, start a
.Xr kldstat 8
and read the information from it.
.It Fl M
Specify the core file for
.Xr kvm 3 .
Implies
.Fl V .
.It Fl N
Specify the system file for
.Xr kvm 3 .
Implies
.Fl V .
.It Fl o
Specify the file for
.Nm
to write or append its output to.
If
.Ar outfile
is a single dash
.Pq Sq Fl ,
standard output is used.
.It Fl s
Do not prepend a (guessed) subdirectory of the module path.
.It Fl V
Instead of reading from standard input, use the
.Xr kvm 3
interface to get the list of modules.
This interface allows for inspecting system crash dumps,
as well as the live system.
The
.Fl M
and
.Fl N
options will be of use if inspecting a crash dump.
Elevated privileges, e.g., those of a superuser,
may be needed to use this option.
.It Fl X
Add
.Ar suffix
to the list of suffixes
.Nm
tries to append to KLD file names.
The default list consists of
.Pa .debug ,
.Pa .symbols ,
and the null suffix.
The null suffix always stays at the list tail, after the suffix added.
Should it be needed in the middle of the list,
a blank suffix can be specified to
.Fl X
instead.
.It Fl x
Clear the list of suffixes
.Nm
tries to append to KLD file names.
Only the null suffix is left in the list.
.El
.Sh EXAMPLES
To add symbol files from the system search path specified by the
.Va kern.module_path
sysctl, the following command can be used:
.Pp
.Dl asf -s `sysctl -n kern.module_path`
.Sh SEE ALSO
.Xr gdb 1 ,
.Xr kvm 3 ,
.Xr kld 4 ,
.Xr kldstat 8 ,
.Xr sysctl 8
.Sh HISTORY
The
.Nm
utility first appeared in
.Fx 5.2 .
.Sh AUTHORS
.An Greg Lehey Aq Mt grog@FreeBSD.org
.Sh BUGS
Module paths are guessed in a rather naive way by default.
It is likely to lag behind the changes to the build tree layout.
Using
.Fl f
is recommended.


NEWS
GNU diffutils NEWS                                    -*- outline -*-

* Noteworthy changes in release 3.2 (2011-09-02) [stable]

** Changes in behavior

  --ignore-file-name-case now applies at the top level too.
  For example, "diff dir inIt" might compare "dir/Init" to "inIt".

** New features

  diff and sdiff have a new option --ignore-trailing-space (-Z).

** Packaging

  The texinfo documentation no longer specifies "front-cover" or "back-cover"
  texts, so that it may now be included in Debian's "main" section.


* Noteworthy changes in release 3.1 (2011-08-10) [stable]

** Bug fixes

  diff no longer reports spurious differences merely because two entries
  in the same directory have names that compare equal in the current
  locale, or compare equal because --ignore-file-name-case was given.

* Noteworthy changes in release 3.0 (2010-05-03) [stable]

** Bug fixes

  diff once again prints the required "\ No newline at end of file" line
  when at least one input lacks a newline-at-EOF and the final hunk plus
  context-length aligns exactly with the end of the newline-lacking file.
  [bug introduced between 2.8.7 and 2.9]

** Changes in behavior

  In context-style diffs, diff prints a portion of a preceding "function"
  line for each hunk, with --show-function-line=RE (-F) or
  --show-c-function (-p).  Now, it trims leading blanks from such lines
  before extracting a prefix.  This is useful especially when a function
  line is so far indented that the name itself would be truncated or not
  included in the limited-width substring that diff appends.

  diff once again reports a difference with the diagnostic
  "Binary files A and B differ" when at least one of the files
  appears to be binary.  From 2.8.4 through diffutils-2.9, it printed
  "Files A and B differ".


* Noteworthy changes in release 2.9 (2010-02-11) [stable]

** New features

  New diff option --suppress-blank-empty.

  Bring back support for `diff -NUM', where NUM is a number,
  even when conforming to POSIX 1003.1-2001.  This change reverts to
  the behavior of GNU diff 2.7 and earlier.  This is a change only
  when conforming to POSIX 1003.1-2001; there is no effect when
  conforming to older POSIX versions.

  This change is in response to decisions taken in the January 2005
  Austin Group standardization meeting.  For more details, please see
  "Utility Syntax Guidelines" in the Minutes of the January 2005
  Meeting <http://www.opengroup.org/austin/docs/austin_239.html>.

  sdiff now understands '1' and '2' as synonyms for 'l' and 'r'.

** Changes in behavior

  sdiff and diff3 now invoke diff, not $(bindir)/diff

** Administrivia

  New discussion and bug-reporting address: bug-diffutils@gnu.org

  updated gnulib support


* Noteworthy changes in release 2.8.7 (2004-04-13) [stable]

  Version 2.8.7 contains no user-visible changes.

User-visible changes in version 2.8.6:

* New diff3 option --strip-trailing-cr.

* With -N and -P, inaccessible empty regular files (the kind of files
  that 'patch' creates to indicate nonexistent backups) are now
  treated as nonexistent when they are in the 'backup' file position.

* If multiple SKIP values are given to cmp, e.g., `cmp -i 10 -i 20',
  cmp now uses the maximal value instead of the last one.

* diff now omits the ".000000000" on hosts that do not support
  fractional time stamps.

Version 2.8.5 was not publicly released.

User-visible changes in version 2.8.4:

* Diff now simply prints "Files A and B differ" instead of "Binary
  files A and B differ".  The message is output if either A or B
  appears to be a binary file, and the old wording was misleading
  because it implied that both files are binary, which is not
  necessarily the case.

User-visible changes in version 2.8.3:

* New locale: en_US.

User-visible changes in version 2.8.2:

* New diff and sdiff option:
  --tabsize=COLUMNS
* If --ignore-space-change or --ignore-all-space is also specified,
  --ignore-blank-lines now considers lines to be empty if they contain
  only white space.
* More platforms now handle multibyte characters correctly when
  excluding files by name (diff -x and -X).
* New locales: hu, pt_BR.

User-visible changes in version 2.8.1:

* Documentation fixes.

User-visible changes in version 2.8:

* cmp and diff now conform to POSIX 1003.1-2001 (IEEE Std 1003.1-2001)
  if the underlying system conforms to POSIX and if the _POSIX2_VERSION
  environment variable is set to 200112.  Conformance removes support
  for `diff -NUM', where NUM is a number.  Use -C NUM or -U NUM instead.
* cmp now supports trailing operands SKIP1 and SKIP2, like BSD cmp.
* cmp -i or --ignore-initial now accepts SKIP1:SKIP2 option value.
* New cmp option: -n or --bytes.
* cmp's old -c or --print-chars option has been renamed;
  use -b or --print-bytes instead.
* cmp now outputs "byte" rather than "char" outside the POSIX locale.
* cmp -l's index column width now adjusts to fit larger (or smaller) files.
* cmp -l -s and cmp -s -l are not allowed.  Use cmp -s or cmp -l instead.
* diff uses ISO 8601 style time stamps for output times (e.g. "2001-11-23
  16:44:36.875702460 -0800") unless in the C or POSIX locale and the
  -c style is specified.
* diff's -I and -F options use the regexp syntax of grep, not of Emacs.
* diff now accepts multiple context arguments, and uses their maximum value.
* New diff and sdiff options:
  -E  --ignore-tab-expansion
  --strip-trailing-cr
* New diff options:
  --from-file=FILE, --to-file=FILE
  --ignore-file-name-case
  --no-ignore-file-name-case
* New diff3 and sdiff option:
  --diff-program=PROGRAM
* The following diff options are still accepted, but are no longer documented.
  They may be withdrawn in future releases.
  -h (omit; it has no effect)
  -H (use --speed-large-files instead)
  -L (use --label instead)
  -P (use --unidirectional-new-file instead)
  --inhibit-hunk-merge (omit; it has no effect)
* Recursive diffs now sort file names according to the LC_COLLATE locale

usr.sbin/asf/asf.c
/*-
 * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
 *
 * Copyright (c) 2002, 2003 Greg Lehey
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * This software is provided by the author ``as is'' and any express
 * or implied warranties, including, but not limited to, the implied
 * warranties of merchantability and fitness for a particular purpose
 * are disclaimed.  In no event shall the author be liable for any
 * direct, indirect, incidental, special, exemplary, or consequential
 * damages (including, but not limited to, procurement of substitute
 * goods or services; loss of use, data, or profits; or business
 * interruption) however caused and on any theory of liability,
 * whether in contract, strict liability, or tort (including
 * negligence or otherwise) arising in any way out of the use of this
 * software, even if advised of the possibility of such damage.
 */
/* $Id: asf.c,v 1.4 2003/05/04 02:55:20 grog Exp grog $ */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: stable/11/usr.sbin/asf/asf.c 330449 2018-03-05 07:26:05Z eadler $");

#include <sys/types.h>
#include <sys/queue.h>
#include <sys/stat.h>
#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <fts.h>
#include <inttypes.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "asf.h"

struct kfile {
    char	       *name;
    caddr_t		addr;
    int			seen;
    STAILQ_ENTRY(kfile)	link;
};

static STAILQ_HEAD(,kfile) kfile_head = STAILQ_HEAD_INITIALIZER(kfile_head);

void
kfile_add(const char *name, caddr_t addr)
{
    struct kfile *kfp;

    if ((kfp = malloc(sizeof(*kfp))) == NULL ||
	(kfp->name = strdup(name)) == NULL)
	    errx(2, "out of memory");
    kfp->addr = addr;
    kfp->seen = 0;
    STAILQ_INSERT_TAIL(&kfile_head, kfp, link);
}

static struct kfile *
kfile_find(const char *name)
{
    struct kfile *kfp;

    STAILQ_FOREACH(kfp, &kfile_head, link)
	if (strcmp(kfp->name, name) == 0)
	    return (kfp);	/* found */

    return (NULL);		/* not found */
}

static int
kfile_allseen(void)
{
    struct kfile *kfp;

    STAILQ_FOREACH(kfp, &kfile_head, link)
	if (!kfp->seen)
	    return (0);	/* at least one unseen */

    return (1);		/* all seen */
}

static int
kfile_empty(void)
{
    return (STAILQ_EMPTY(&kfile_head));
}

/*
 * Take a blank separated list of tokens and turn it into a list of
 * individual nul-delimited strings.  Build a list of pointers at
 * token, which must have enough space for the tokens.  Return the
 * number of tokens, or -1 on error (typically a missing string
 * delimiter).
 */
int
tokenize(char *cptr, char *token[], int maxtoken)
{
    char delim;				/* delimiter to search for */
    int tokennr;			/* index of this token */

    for (tokennr = 0; tokennr < maxtoken;) {
	while (isspace(*cptr))
	    cptr++;			/* skip initial white space */
	if ((*cptr == '\0') || (*cptr == '\n')
	    || (*cptr == '#'))		/* end of line */
	    return tokennr;		/* return number of tokens found */
	delim = *cptr;
	token[tokennr] = cptr;		/* point to it */
	tokennr++;			/* one more */
	if (tokennr == maxtoken)	/* run off the end? */
	    return tokennr;
	if ((delim == '\'') || (delim == '"')) { /* delimitered */
	    for (;;) {
		cptr++;
		if ((*cptr == delim)
		    && (cptr[-1] != '\\')) { /* found the partner */
		    cptr++;		/* move on past */
		    if (!isspace(*cptr)) /* no space after closing quote */
			return -1;
		    *cptr++ = '\0';	/* delimit */
		} else if ((*cptr == '\0')
		    || (*cptr == '\n'))	/* end of line */
		    return -1;
	    }
	} else {			/* not quoted */
	    while ((*cptr != '\0') && (!isspace(*cptr)) && (*cptr != '\n'))
		cptr++;
	    if (*cptr != '\0')		/* not end of the line, */
		*cptr++ = '\0';		/* delimit and move to the next */
	}
    }
    return maxtoken;			/* can't get here */
}

static void
doobj(const char *path, caddr_t addr, FILE *out)
{
    uintmax_t	base = (uintptr_t)addr;
    uintmax_t	textaddr = 0;
    uintmax_t	dataaddr = 0;
    uintmax_t	bssaddr = 0;
    uintmax_t  *up;
    int		octokens;
    char       *octoken[MAXTOKEN];
    char	ocbuf[LINE_MAX + PATH_MAX];
    FILE       *objcopy;

    snprintf(ocbuf, sizeof(ocbuf),
	     "/usr/bin/objdump --section-headers %s", path);
    if ((objcopy = popen(ocbuf, "r")) == NULL)
	err(2, "can't start %s", ocbuf);
    while (fgets(ocbuf, sizeof(ocbuf), objcopy)) {
	octokens = tokenize(ocbuf, octoken, MAXTOKEN);
	if (octokens <= 1)
	    continue;
	up = NULL;
	if (strcmp(octoken[1], ".text") == 0)
	    up = &textaddr;
	else if (strcmp(octoken[1], ".data") == 0)
	    up = &dataaddr;
	else if (strcmp(octoken[1], ".bss") == 0)
	    up = &bssaddr;
	if (up == NULL)
	    continue;
	*up = strtoumax(octoken[3], NULL, 16) + base;
    }
    if (textaddr) {	/* we must have a text address */
	fprintf(out, "add-symbol-file %s 0x%jx", path, textaddr);
	if (dataaddr)
	    fprintf(out, " -s .data 0x%jx", dataaddr);
	if (bssaddr)
	    fprintf(out, " -s .bss 0x%jx", bssaddr);
	fprintf(out, "\n");
    }
}

static void
findmodules(char *path_argv[], const char *sfx[], FILE *out)
{
    char	       *p;
    FTS		       *fts;
    FTSENT	       *ftsent;
    struct kfile       *kfp;
    int			i;
    int			sl;

    /* Have to fts once per suffix to find preferred suffixes first */
    do {
	sl = *sfx ? strlen(*sfx) : 0;	/* current suffix length */
	fts = fts_open(path_argv, FTS_PHYSICAL | FTS_NOCHDIR, NULL);
	if (fts == NULL)
	    err(2, "can't begin traversing module path");
	while ((ftsent = fts_read(fts)) != NULL) {
	    if (ftsent->fts_info == FTS_DNR ||
		ftsent->fts_info == FTS_ERR ||
		ftsent->fts_info == FTS_NS) {
		    errno = ftsent->fts_errno;
		    err(2, "error while traversing path %s", ftsent->fts_path);
	    }
	    if (ftsent->fts_info != FTS_F)
		continue;			/* not a plain file */

	    if (sl > 0) {
		/* non-blank suffix; see if file name has it */
		i = ftsent->fts_namelen - sl;
		if (i <= 0 || strcmp(ftsent->fts_name + i, *sfx) != 0)
		    continue;		/* no such suffix */
		if ((p = strdup(ftsent->fts_name)) == NULL)
		    errx(2, "out of memory");
		p[i] = '\0';		/* remove suffix in the copy */
		kfp = kfile_find(p);
		free(p);
	    } else
		kfp = kfile_find(ftsent->fts_name);

	    if (kfp && !kfp->seen) {
		doobj(ftsent->fts_path, kfp->addr, out);
		kfp->seen = 1;
		/* Optimization: stop fts as soon as seen all loaded modules */
		if (kfile_allseen()) {
		    fts_close(fts);
		    return;
		}
	    }
	}
	if (ftsent == NULL && errno != 0)
	    err(2, "couldn't complete traversing module path");
	fts_close(fts);
    } while (*sfx++);
}

static void
usage(const char *myname)
{
    fprintf(stderr,
	"Usage:\n"
	"%s [-afKksVx] [-M core] [-N system] [-o outfile] [-X suffix]\n"
	"%*s [modules-path [outfile]]\n\n"
	"\t-a\tappend to outfile\n"
	"\t-f\tfind the module in any subdirectory of modules-path\n"
	"\t-K\tuse kld(2) to get the list of modules\n"
	"\t-k\ttake input from kldstat(8)\n"
	"\t-M\tspecify core name for kvm(3)\n"
	"\t-N\tspecify system name for kvm(3)\n"
	"\t-o\tuse outfile instead of \".asf\"\n"
	"\t-s\tdon't prepend subdir for module path\n"
	"\t-V\tuse kvm(3) to get the list of modules\n"
	"\t-X\tappend suffix to list of possible module file name suffixes\n"
	"\t-x\tclear list of possible module file name suffixes\n",
	myname, (int)strlen(myname), "");
    exit(2);
}

#define	MAXPATHS	15
#define	MAXSUFFIXES	15

/* KLD file names end in this */
static int	   nsuffixes = 2;
static const char *suffixes[MAXSUFFIXES + 1] = {
    ".debug",
    ".symbols",
    NULL
};

int
main(int argc, char *argv[])
{
    char basename[PATH_MAX];
    char path[PATH_MAX];
    char *modules_argv[MAXPATHS + 1];
    char *copy, *p;
    char **ap;
    const char *filemode = "w";		/* mode for outfile */
    const char *modules_path = "modules"; /* path to kernel build directory */
    const char *outfile = ".asf";	/* and where to write the output */
    const char *corefile = NULL;	/* for kvm(3) */
    const char *sysfile = NULL;		/* for kvm(3) */
    const char **sfx;
    struct kfile *kfp;
    struct stat st;
    FILE *out;				/* output file */
    int dofind = 0;
    int dokld = 0;
    int dokvm = 0;
    int nosubdir = 0;
    int runprog = 0;
    int i;
    const int sl = strlen(KLDSUFFIX);

    while ((i = getopt(argc, argv, "afKkM:N:o:sVX:x")) != -1)
	switch (i) {
	case 'a':
	    filemode = "a";	/* append to outfile */
	    break;
	case 'f':
	    dofind = 1;		/* find .ko (recursively) */
	    break;
	case 'K':
	    dokld = 1;		/* use kld(2) interface */
	    break;
	case 'k':
	    runprog = 1;	/* get input from kldstat(8) */
	    break;
	case 'M':
	    corefile = optarg;	/* core file for kvm(3) */
	    break;
	case 'N':
	    sysfile = optarg;	/* system file (kernel) for kvm(3) */
	    break;
	case 'o':
	    outfile = optarg;	/* output file name */
	    break;
	case 's':
	    nosubdir = 1;	/* don't descend into subdirs */
	    break;
	case 'V':
	    dokvm = 1;		/* use kvm(3) interface */
	    break;
	case 'X':
	    if (nsuffixes >= MAXSUFFIXES)
		errx(2, "only %d suffixes can be specified", MAXSUFFIXES);
	    suffixes[nsuffixes++] = optarg;
	    suffixes[nsuffixes] = NULL;
	    break;
	case 'x':
	    nsuffixes = 0;
	    suffixes[0] = NULL;
	    break;
	default:
	    usage(argv[0]);
	}

    argc -= optind;
    argv += optind;

    if (argc > 0) {
	modules_path = argv[0];
	argc--, argv++;
    }
    if (argc > 0) {
	outfile = argv[0];
	argc--, argv++;
    }
    if (argc > 0)
	usage(argv[0]);

    if (strcmp(outfile, "-") == 0)
	out = stdout;
    else
	if ((out = fopen(outfile, filemode)) == NULL)
	    err(2, "can't open output file %s", outfile);

    if (dokvm || corefile || sysfile) {
	if (dokld || runprog)
	    warnx("using kvm(3) instead");
	asf_kvm(sysfile, corefile);
    } else if (dokld) {
	if (runprog)
	    warnx("using kld(2) instead");
	asf_kld();
    } else
	asf_prog(runprog);

    /* Avoid long operations like module tree traversal when nothing to do */
    if (kfile_empty()) {
	warnx("no kernel modules loaded");
	return (0);
    }

    if ((copy = strdup(modules_path)) == NULL)
	errx(2, "out of memory");
    for (
	ap = modules_argv, p = copy;
	(*ap = strsep(&p, ";")) != NULL && ap < &modules_argv[MAXPATHS];
	ap++
    );
    if (*ap)
	errx(2, "only %d module path elements can be specified", MAXPATHS);

    if (!dofind)
	STAILQ_FOREACH(kfp, &kfile_head, link) {
	    for (ap = modules_argv; *ap; ap++) {
		if (!nosubdir) {
		    /* prepare basename of KLD, w/o suffix */
		    strlcpy(basename, kfp->name, sizeof(basename) - 1);
		    i = strlen(basename);
		    if (i > sl && strcmp(basename + i - sl, KLDSUFFIX) == 0)
			i -= sl;
		    basename[i] = '/';
		    basename[i + 1] = '\0';
		}
		for (sfx = suffixes;; sfx++) {
		    snprintf(path, sizeof(path),
			     "%s/%s%s%s",
			     *ap,
			     nosubdir ? "" : basename,
			     kfp->name,
			     *sfx ? *sfx : "");
		    if (stat(path, &st) == 0) {
			doobj(path, kfp->addr, out);
			goto found;
		    }
		    if (*sfx == NULL)
			break;
		}
	    }
	    warnx("module %s not found in search path", kfp->name);
found:
	    ;
	}
    else
    	findmodules(modules_argv, suffixes, out);

    free(copy);
    return (0);
}
